life_expectancy/cleaning.py

```diff
@@ -0,0 +1,188 @@
+
+
+def remove_non_numeric_characters_from_column(data: pd.DataFrame, column_name: str) -> pd.Series:
+    """Removes all non-numeric characters from a given column."""
+    return data[column_name].apply(lambda x: re.sub(r'[^\d.]', '', str(x)))
+
+
+def convert_column_to_numeric(data: pd.DataFrame, column_name: str) -> pd.Series:
+    """Converts a given column to numeric and converts non-numerical values to NaN."""
+    return pd.to_numeric(data[column_name], errors='coerce')
+
+
+def remove_nan_by_column_subset(data: pd.DataFrame, column_subset: list[str]) -> pd.DataFrame:
+    """
+    Removes rows from a DataFrame based on the rows with NaN of a given subset 
+    of columns.
+    """
+    return data.dropna(subset=column_subset).reset_index(drop=True)
+
+
+def clean_value_column(data: pd.DataFrame) -> pd.DataFrame:
+    """Processes the 'value' column by performing a set of operations."""
+    data['value'] = remove_non_numeric_characters_from_column(data, 'value')
+    data['value'] = convert_column_to_numeric(data, 'value')
+    filtered_data = remove_nan_by_column_subset(data, ['value'])
+    filtered_data['value'] = change_column_type(filtered_data, 'value', float)
+    
+    return filtered_data
+
+
+def filter_data_by_region(data: pd.DataFrame, region: str) -> pd.DataFrame:
+    """
+    Filters the DataFrame by a given region.
+    Raises a RegionNotFoundError if the region does not exist in the data.
+
+    Example:
+    >>> filtered_data = filter_data_by_region(data, 'PT')
+    """
+    filtered_data = data[data['region'] == region]
+
+    if filtered_data.empty:
+        valid_regions = data['region'].unique().tolist()
+        raise RegionNotFoundError(region, valid_regions)
+
+    return filtered_data.reset_index(drop=True)
+
+
+def clean_data(data: pd.DataFrame, region: str) -> pd.DataFrame:
+    """
+    Cleans the input data by transforming and filtering the original DataFrame 
+    and processing its columns.
+    """
+    transformed_data = transform_data(data)
+    transformed_data_year_cleaned = clean_year_column(transformed_data)
+    transformed_data_value_cleaned = clean_value_column(transformed_data_year_cleaned)
+    final_data = filter_data_by_region(transformed_data_value_cleaned, region)
+
+    return final_data
+
+
+def save_data(data: pd.DataFrame, region: str) -> None:
+    """Exports the DataFrame to a text file."""
+    base_dir = get_base_directory()
+    output_file_path = os.path.join(base_dir, 'data', f'{region.lower()}_life_expectancy.csv')
+    data.to_csv(output_file_path, index=False)
+
+
+def run_all_functions(region: str) -> None:
+    """Makes the functions call in order. This is the main function."""
+    try:
+        raw_data = load_data()
+        if raw_data is not None:
+            cleaned_data = clean_data(raw_data, region)
+            save_data(cleaned_data, region)
+    except RegionNotFoundError as e:
+        print(e)
+
+
+if __name__ == "__main__":  # pragma: no cover
```

Comment:
I expected to these functions to be called in a `__main__.py` file and perhaps the `cleaning.py` to be part of a package, this package could later have other files that would treat the data. Could you please elaborate about the code structure?

Reply:
I am not sure if I fully understood your question. However, following the standard project structure, cleaning.py would not be the main script. I would have another script called main.py, as you suggested, which would call the functions from all other necessary scripts. The statement if __name__ == "__main__": is only there for debugging purposes. If I want to test cleaning.py, instead of running main.py to run cleaning.py, I run cleaning.py directly. The statement if __name__ == "__main__": will only be active when you run the script directly. In this case, __name__ == "__main__".

Reply:
You are right, my comment is a little confusing. Let me re-iterate: 

What I meant is, you could have the cleaning.py file as a module, while having a `__main__.py` file which imports functions from this module. 

Imagine if the project expands and you have more stages, data preprocessing, feature engineering, modeling, validation and persistence of the model in some storage location (eg. database)

The `__main__.py` could be responsible for running all the stages end-to-end, with each stage having its own module/sub-package

In essence, I am suggesting a structure like this: 

```
your_project/
├── data/
│   └── eu_life_expectancy_raw.tsv
├── cleaning.py
├── __main__.py  
```

with `__main__.py` looking something like this:

```
import argparse
from cleaning import run_all_functions


def main():
    parser = argparse.ArgumentParser(description="Process life expectancy data by region.")
    parser.add_argument('--region', type=str, default='PT',
                        help="The region to filter the data by. Default is 'PT'.")
    args = parser.parse_args()
    run_all_functions(args.region)


if __name__ == "__main__":
    main()
```

`run_all_functions` could instead be called run_cleaning_data. 

This will help you maintain a more modular structure, and will help you have a single __main__.py file that users interact with, and will have separation of concerns, where the program logic is encoded into `__main__.py` while other stages of the pipeline (cleaning, feature_engineering, etc.) have their own modules. Another important addition to this structure, is that in the __main__.py file, you can introduce loggers, which give you hints on what's happening in the full program stages (eg. Cleaning concluded in x seconds, feature_engineering concluded in x seconds, database connection failed etc.).

It is usually much easier to implement loggers in the `__main__.py` file, and not everywhere around the program. 

I think this recommendation I am making makes sense when you have more functionalities or stages in the pipeline, where each step needs to run sequentially, it eventually makes sense to group it in a __main__.py file which connects the output of each stage to the next one.

Reply:
Yes, of course. Totally agree with you and that's the best practice. My question is, I am not sure if it is supposed to do that right now or we will be asked to do so in later assignments. We can ask Bruno this :)

Reply:
Actually, I think I am going to do that anyway. I was checking this week's assignment, and I think we should do that. In any case, I was feeling the need to separate stuff ;p Thank you!